### PR TITLE
#73 Enable individual components import for tree shaking

### PR DESCRIPTION
**Pull request by nizniz187**
Individual component import is fullfilled by the following approaches:

1. **Define custom elements within each component JS as side effects.**
    Some components with dependencies may need addittional import & type check, like `Form` & `Option`.
2. **Import & re-export all compoment modules in `w-components/index.js`.**
    For later module import use and tree shaking (yes, we are actually doing tree shaking for bundling, not code splitting.)
3. **Add `sideEffects: false` in `w-components/packages.json`.**
    So that webpack will skip the side effects and bundle modules only by imports & exports. See: https://webpack.js.org/guides/tree-shaking/#mark-the-file-as-side-effect-free.
4. **Set webpack compile mode to `production`.**
    By default webpack's tree-shaking mechanism works on `production` mode without no additional setting. See: https://stackoverflow.com/a/56351070.

So that if we import specific components in `index.js`:
```javascript
import wc, { Button } from '@padax/w-components';

export default wc;
export { Button };
```
We can see nothing is working, except for `Button`.
image
And the bundle size is decreasing to 130KiB:
image

---

### Issues
Unsurprisingly, there is an issue about loading sequence, as this was the reason we chose to define custom elements all in once with global config & theme setting after components imported.
For now we can remain the `wc.init` way for global config & theme setting, and leave this to be figured out in #94. But for `SPA basename` config, we may need to find a new way for this tree shaking solution. 
**So @cwpeng please confirm that if we can set the `basename` by passing it as an attribute.**

### CODE REVIEW

File: packages/w-components/components/form/Form.js [hunk 1]
```diff
@@ -183,7 +185,7 @@ class Form extends WComponent{
     const radios = Array.from(this.querySelectorAll(`${getWTagName('radio')}[name='${radio.name}']`));
     // Compare each w-radio in the same group, uncheck all which are not the current event target.
     radios.forEach(r => {
-      if(!r.equals(radio)) {
+      if(r instanceof Radio && !r.equals(radio)) {
```
Comment (nizniz187) [inline packages/w-components/components/form/Form.js:188]: Additional type check.

File: packages/w-components/components/spa/SPAPage.js [hunk 2]
```diff
@@ -82,4 +83,9 @@ class SPAPage extends WComponent{
   }
 }
 SPAPage.prototype.stylesheet=stylesheet;
+
+const BASE_NAME = window.location.pathname.substring(0, window.location.pathname.lastIndexOf("/"));
```
Comment (nizniz187) [inline packages/w-components/components/spa/SPAPage.js:87]: Add this basename setting temporary for avoiding exception caused by wconfig not set.

File: packages/w-components/util/DOM.js [hunk 3]
```diff
@@ -65,6 +65,15 @@ const DOM={
       }
     }
     return element;
+  },
+  defineCustomElement(component) {
+    const tag = `${customElementPrefix}-${component.tagName}`;
+    if(!customElements.get(tag)) {
+      customElements.define(tag, component);
+    }
   }
```
Comment (nizniz187) [inline packages/w-components/util/DOM.js:74]: Include defineCustomElement function in `DOM.js`.

**Review by cwpeng — APPROVED** [review]
OK.

**Pull request merged**